### ISSUE TITLE
Early Session Invalidation

### ISSUE THREAD

**Issue by sirwolfgang** [hive-xyz/hive#20]
In the event that the data stored by a session is changed in a breaking way, like the addition of a critical attribute, there needs to be a way to invalidate all current tokens.

**Comment by davidrivera** [comment]
one possible way to invalidate all tokens could be to just to provide a custom validation against the issued_at claim of the JWT, anything issued before the date at which you want to invalidate all tokens, on tokens that would otherwise be considered valid would now be considered invalid, everything after said date would just be run through the normal checks. This is probably the easiest  and most straight forward approach.

tl;dr
Get Token -> Validate -> Is Valid -> Yes -> Is invalidate all date set -> Yes -> Run extra validation step

**Comment by sirwolfgang** [comment]
Ya, I was thinking about something along those lines. Do you think that logic should be application side, or rolled into [keeper](https://github.com/hive-xyz/keeper) as a feature?

**Comment by davidrivera** [comment]
Yea i'm not sure where is a good place to put this, It would def be nice as a feature in keeper but we need to figure out a way to configure a date throughout multiple requests. The date could possibly live in either the db, redis, or an env variable, possibly even in a text file. Maybe the storage engine for this should be configurable? I would really like to do something like `Keeper.invalidate_all_tokens!` or something along those lines.

**Comment by sirwolfgang** [comment]
`Keeper.invalidate_all_tokens!` is tempting, but I can't really think of a situation where having the ability to runtime invalidate would be that useful. The biggest case I see is version shifts, which already require a redeploy; making an ENV pretty easy. The secondary case I see is in the event of a security issues, but should result in a secret change which would invalidate anyway.

If it was a runtime, it would have to be data store backed, which would mean that it should go in redis with the blacklist. If you were rotating out a version change, you would need to wait until full rotation before a trigger. This should be implemented as part of keeper. 

If it was deploy time, it could be deployed out with code as you rotate out your cluster. Making new connections kill and generate new sessions, while current connections could roll over somewhat gracefully. This should be implemented as a config into keeper.

@davidrivera Can you think of a case where via logic, or API call, it would be useful to programmatically invalidate all tokens?

**Comment by sirwolfgang** [comment]
Moving to hive-xyz/keeper#11
